app/uk/gov/hmrc/gform/instructions/InstructionsRenderingService.scala

```diff
@@ -210,41 +214,82 @@ class InstructionsRenderingService(
         }
       }).getOrElse(List.empty)
 
-    val formModel = formModelOptics.formModelVisibilityOptics.formModel
-    formModel.singletonsBySource
+    def addToListRender(addToList: Section.AddToList): Html = {
+      val repeaters: List[Repeater[Visibility]] = formModel.repeaters(addToList.id)
+      val recordTable: List[SmartString] = repeaters.map(_.expandedDescription)
+      val value = recordTable.map(_.value()).mkString("</br>")
+
+      val slr: SummaryListRow = summaryListRow(
+        addToList.title.value(),
+        value,
+        None,
+        "",
+        "",
+        "",
+        Nil
+      )
+
+      new govukSummaryList()(SummaryList(slr :: Nil, "govuk-!-margin-bottom-5"))
+    }
+
+    def instructionOrder(section: Section) =
+      section
+        .fold(_.page.instruction.flatMap(_.order))(_.page.instruction.flatMap(_.order))(_.instruction.flatMap(_.order))
+        .getOrElse(Integer.MAX_VALUE)
+
+    formModel.pagesWithIndex
+      .groupBy {
+        case (pageModel, _) => pageModel.source
+      }
+      .toList
       .sortBy {
-        case (section, _) =>
-          section
-            .fold(_.page.instruction.flatMap(_.order))(_.page.instruction.flatMap(_.order))(
-              _.instruction.flatMap(_.order))
-            .getOrElse(Integer.MAX_VALUE)
+        case (section, _) => instructionOrder(section)
       }
       .flatMap {
-        case (section, pageModels) =>
-          val header =
-            section.fold(_ => List.empty[HtmlFormat.Appendable])(_ => List.empty[HtmlFormat.Appendable])(addToList =>
-              addToList.instruction.map(i => begin_section(i.name)).toList)
-          header ++ pageModels.flatMap(pModel => {
-            pModel.fold(renderHtmls(_, formModel.pagesMap(pModel)))(_ => Nil)
-          })
+        case (a: AddToList, pagesWithSectionNumber) =>
+          List(addToListRender(a)) ++ pagesWithSectionNumber
+            .collect {
+              case (singleton: Singleton[Visibility], sectionNumber) =>
+                (formModel.repeaterForSingleton(singleton, sectionNumber).get, (singleton, sectionNumber))
```

Comment:
Have you consider using existing
```
def repeaterFor(addToListId: AddToListId): Option[Repeater[A]]
```
here ?

Also please don't call `.get` on an `Option`

Reply:
Thanks @josef-vlach. I have removed the usage of Option.get

I found the logic in SummaryRenderingService, using `repeaterFor(index: Int, addToListId: AddToListId)` a bit hard to follow, thats why resorted to using grouping. To use `repeaterFor(index: Int, addToListId: AddToListId)`, I need to remember the previous repeaters index, which I don't have here. Do you see an issue with this approach?